File: app/src/main/java/com/example/bottlenex/LocationDetailsActivity.java
package com.example.bottlenex;

import android.os.Bundle;
import android.view.Gravity;
import android.widget.Button;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import androidx.core.content.ContextCompat;

public class LocationDetailsActivity extends AppCompatActivity {
    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_location_details);

        // Set up CIF keyboard
        setupKeyboard();

        // Get data from intent
        String locationName = getIntent().getStringExtra("location_name");
        if (locationName == null) locationName = "CLEMENTI";
        ((TextView) findViewById(R.id.tvLocationTitle)).setText(locationName);
    }

    private void setupKeyboard() {
        String[][] rows = {
                {"q", "w", "e", "r", "t", "y", "u", "i", "o", "p"},
                {"a", "s", "d", "f", "g", "h", "j", "k", "l"},
                {"⭕", "z", "x", "c", "v", "b", "n", "m", "💶"}
        };
        LinearLayout keyboardContainer = findViewById(R.id.keyboardContainer);
        for (String[] rowKeys : rows) {
            LinearLayout rowLayout = new LinearLayout(this);
            rowLayout.setLayoutParams(new LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
            ));
            rowLayout.setGravity(Gravity.CENTER);
            for (final String key : rowKeys) {
                Button button = new Button(this);
                button.setText(key);
                LinearLayout.LayoutParams params = new LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.WRAP_CONTENT,
                        LinearLayout.LayoutParams.WRAP_CONTENT
                );
                int margin = dpToPx(4);
                params.setMargins(margin, margin, margin, margin);
                button.setLayoutParams(params);
                button.setBackground(ContextCompat.getDrawable(this, R.drawable.keyboard_key_bg));
                button.setOnClickListener(v -> onKeyPressed(key));
                rowLayout.addView(button);
            }
            keyboardContainer.addView(rowLayout);
        }
    }

    private int dpToPx(int dp) {
        return (int) (dp * getResources().getDisplayMetrics().density);
    }

    private void onKeyPressed(String key) {
        Toast.makeText(this, "Pressed: " + key, Toast.LENGTH_SHORT).show();
    }
} 

File: app/src/main/java/com/example/bottlenex/LocationDetailsActivity.kt
package com.example.bottlenex

import android.os.Bundle
import android.view.Gravity
import android.widget.Button
import android.widget.LinearLayout
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat

class LocationDetailsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_location_details)

        // Set up CIF keyboard
        setupKeyboard()

        // Get data from intent
        val locationName = intent.getStringExtra("location_name") ?: "CLEMENTI"
        findViewById<TextView>(R.id.tvLocationTitle).text = locationName
    }

    private fun setupKeyboard() {
        val rows = listOf(
            listOf("q", "w", "e", "r", "t", "y", "u", "i", "o", "p"),
            listOf("a", "s", "d", "f", "g", "h", "j", "k", "l"),
            listOf("⭕", "z", "x", "c", "v", "b", "n", "m", "💶")
        )

        val keyboardContainer = findViewById<LinearLayout>(R.id.keyboardContainer)

        rows.forEach { rowKeys ->
            val rowLayout = LinearLayout(this).apply {
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
                gravity = Gravity.CENTER
            }

            rowKeys.forEach { key ->
                val button = Button(this).apply {
                    text = key
                    layoutParams = LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.WRAP_CONTENT,
                        LinearLayout.LayoutParams.WRAP_CONTENT
                    ).apply {
                        setMargins(4.dpToPx(), 4.dpToPx(), 4.dpToPx(), 4.dpToPx())
                    }
                    background = ContextCompat.getDrawable(this@LocationDetailsActivity, R.drawable.keyboard_key_bg)
                    setOnClickListener { onKeyPressed(key) }
                }
                rowLayout.addView(button)
            }

            keyboardContainer.addView(rowLayout)
        }
    }

    private fun Int.dpToPx(): Int = (this * resources.displayMetrics.density).toInt()

    private fun onKeyPressed(key: String) {
        // Handle keyboard presses
        Toast.makeText(this, "Pressed: $key", Toast.LENGTH_SHORT).show()
    }
}
